coralproject/talk
# [CORL-2608]: Virtualized rendering and next unseen updates

## What does this PR do?

These changes implement virtualized rendering using Virtuoso in the stream. They update to load 20 comments initially, and then all of the rest of the stream comments in the background. For this, loading states were added to the `Load all comments` button that will be seen if anyone clicks on the button or tries to navigate to next unread before all comments are loaded in. Loading states were also added for new comments and new replies buttons.

They also update next unseen comment to work with virtualized rendering. Unmark all now unmarks all comments currently on a stream, not just those that are currently displayed/loaded.

In addition, there are updates to ignored comments and tombstones. Ignored comments are now always shown in the stream. Tombstone styles have been updated.

Media (twitter, youtube) in comments now have an `Expand` option to view them fully if they are over a certain height.

In alternate oldest view, if you click on the `Add a comment` button, it scrolls you down to the `Post a comment` box below all comments.

## What changes to the GraphQL/Database Schema does this PR introduce?

`allChildComments` is added to comments to be used by next unseen and unmark all
`markAllAsSeen` is added as an option to the `markCommentsAsSeen` mutation to support unmarking all comments

## Does this PR introduce any new environment variables or feature flags? 

no

## If any indexes were added, were they added to `INDEXES.md`?

n/a

## How do I test this PR?

Test out the stream in general (adding comments, replying, ignoring users, rejecting comments, etc.).

To test out new loading states, wrap lines 178-180 and 203-210 in `AllCommentsTabContainer` in a setTimeout of something like 8000 or 10000, to mimic 8-10 seconds of loading time. You can test the new replies button loading state by wrapping 285-293 in `ReplyListContainer` in a similar setTimeout. Try to hit the `Load all comments` button before comments are loaded, and also try to navigate with Z key / mobile unread button before comments are loaded and see that you are shown a loading state.

Test out next unseen comment traversal, both via Z key and via the `Next unread` button in the mobile toolbar. Test in combination with ignored users, rejected comments, deleted comments. Test out looping around, new comment and reply subscriptions, unmarking all, when the next unseen comment is far down the stream, etc.

Ignore a user who has left comments. See the ignored comments changes (message/tombstone always shown in the stream). Check out the other tombstones displayed in the stream (moderate a comment and reject it in the stream, delete a comment).

Test out alternate oldest view by sorting by oldest and enabling the feature flag. See that clicking `Add a comment` takes you to the `Post a comment` box below all comments.

Test out expanding, showing, and hiding media (twitter, youtube) in comments in the stream.
 
## How do we deploy this PR?

n/a
